LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/DeviceLoaderSynchronizer.cs

```diff
@@ -146,60 +117,70 @@ private async Task Load()
+                    using var scope = this.logger.BeginDeviceScope(foundDevice.DevEUI);
+                    // Only create devices that does not exist in the cache
+                    if (!this.loraDeviceCache.TryGetByDevEui(foundDevice.DevEUI, out var cachedDevice))
+                    {
+                        initTasks.Add(this.deviceFactory.CreateAndRegisterAsync(foundDevice, CancellationToken.None));
+                    }
+                    else
                     {
-                        using var scope = this.logger.BeginDeviceScope(foundDevice.DevEUI);
-                        var loRaDevice = this.deviceFactory.Create(foundDevice);
-                        initTasks.Add(InitializeDeviceAsync(loRaDevice));
+                        if (string.IsNullOrEmpty(cachedDevice.DevAddr))
+                        {
+                            // device in cache from a previous join that we didn't complete
+                            // (lost race with another gw) - refresh the twins now and keep it
+                            // in the cache
+                            refreshTasks ??= new List<Task<bool>>();
+                            refreshTasks.Add(cachedDevice.InitializeAsync(this.configuration, CancellationToken.None));
+                            this.logger.LogDebug("refreshing device to fetch DevAddr");
+                        }
                     }
                 }
 
                 try
                 {
                     _ = await Task.WhenAll(initTasks);
+                    if (refreshTasks != null)
+                    {
+                        _ = await Task.WhenAll(refreshTasks);
+                    }
                 }
                 catch (LoRaProcessingException ex) when (ex.ErrorCode == LoRaProcessingErrorCode.DeviceInitializationFailed
                                                          && ExceptionFilterUtility.True(() => this.logger.LogError($"one or more device initializations failed: {ex}")))
                 {
                     // continue
+                    HasLoadingDeviceError = true;
                 }
-            }
 
-            var createdDevices = new List<LoRaDevice>();
-            if (initTasks.Count > 0)
-            {
-                foreach (var deviceTask in initTasks)
+                if (initTasks.Count > 0)
                 {
-                    if (deviceTask.IsCompletedSuccessfully)
+                    foreach (var deviceTask in initTasks)
                     {
-                        var device = await deviceTask;
-
-                        if (device != null)
+                        if (deviceTask.IsCompletedSuccessfully)
                         {
+                            var device = await deviceTask;
+                            // run initializers
+                            InitializeDevice(device);
                             createdDevices.Add(device);
                         }
```

Comment:
It looks like this createddevice list is only used to count the number of device created or am I missing something? Couldn't we replace this by a counter if yes?

Reply:
That is true. I consider this a small change, and include it in this scope.

Reply:
941095378e1afdccfa4f7d88e44d45a739ee7376